Various Google Pay button fixes (3477)

stracker-phil opened the pull request:
### Description

This PR (a) addresses several minor issues with the Google Pay button, and (b) provides a cleaner and more structured JS base class for the payment button.

### Addressed issues

#### 1. Button disappears when changing shipping method

- On the classic cart, the Google Pay button disappeared instantly when changing the shipping method
- On classic checkout, the Google Pay button eventually disappeared in the _Google Pay Gateway_ when changing the shipping method multiple times

<details>
<summary><strong>📸 Video: Classic Cart</strong></summary>

https://github.com/user-attachments/assets/f510f1f4-5ab0-4c32-ba44-b95d8fe3fbb1

</details>
<details>
<summary><strong>📸 Video: Classic Checkout</strong></summary>

https://github.com/user-attachments/assets/12e9d16a-9d50-4178-b57f-41a38810d985

</details>

#### 2. Block Checkout created duplicate Google Pay instances

- On the block checkout page, two instances of the Google Pay button were created. The reason for this was a duplicate execution context created by some blocks, and resulted in two payment buttons being injected 
  - Already [observed in PCP-2689](https://inpsyde.atlassian.net/browse/PCP-2689?focusedCommentId=300684) (internal link)

<details>
<summary><strong>📸 Screenshot: Payment button logs on block checkout</strong></summary>

<img width="1084" alt="2024-08-11_13-32-36" src="https://github.com/user-attachments/assets/2994883f-c602-41e9-bf67-69df60c3c353">

</details>

#### 3. Fixed Mini-Cart on product page

- When using Google Pay inside the Mini-Cart while on a product page, the mini-cart now charges for the cart contents, and not the visible product.

<details>
<summary><strong>📸 Screenshot: Mini-Cart on product page</strong></summary>

<img width="1084" alt="2024-08-11_14-12-17" src="https://github.com/user-attachments/assets/e6b9a9f9-28ff-4969-a5f7-c32cd21bb27f">

</details>

#### 4. Button previews in settings

- When changing the button shape in the "Standards Payments" tab, the Google Pay button did not apply the new shape before the page was saved. Now: The Google Pay button updates, like all other PayPal buttons, in real time.

<details>
<summary><strong>📸 Video: Button Preview</strong></summary>

https://github.com/user-attachments/assets/b5906817-e978-4f4c-98f2-6da9e40fce8d

(Note how the Apple Pay preview gets "stuck" once it's changed to "pill" shape once)

</details>

#### 5. Button layout in front-end

- The white Google Pay button correctly shows the dark outline

<details>
<summary><strong>📸 Screenshot: White button in classic checkout</strong></summary>

<img width="644" alt="2024-08-11_14-00-51" src="https://github.com/user-attachments/assets/d16ef6dd-f570-4333-9c80-d8f4b0b3570a">

</details>

### Code structure

- Create a new base class `PaymentButton` 
  - Contains method-independent code, i.e. logic that is also relevant for Apple Pay and other buttons
- Use of JS events to show or hide payment button elements
  - See `PaymentButtonHelpers.js` for [new helpers](https://github.com/woocommerce/woocommerce-paypal-payments/pull/2496/files#diff-f6dc1bc70b9cc91d045ba7829a2a704157610a97d33e041263bdde75a4f482fb).
  - New helpers are used in `CheckoutBootstrap.js` [here](https://github.com/woocommerce/woocommerce-paypal-payments/pull/2496/files#diff-c8cfafba052eed1e40e090694be266dc7090b61185cc87302c0a7240c81d3222R182-R188) and [here](https://github.com/woocommerce/woocommerce-paypal-payments/pull/2496/files#diff-c8cfafba052eed1e40e090694be266dc7090b61185cc87302c0a7240c81d3222R254-R257).
  - Many other small improvements, like the centralized `GooglepayButton.getWrappers()` helper method to pluck CSS IDs from provided config objects, or many new read-only-properties in the `GooglepayButton` instead of public attributes

stracker-phil committed:
🐛 Prevent duplicate payment button instances

stracker-phil committed:
✨ Improve debug-logging for GooglePayButton events

Implement same logic as we use for the ApplePayButton

stracker-phil committed:
💡 Improve/adjust debug logging

stracker-phil committed:
♻️ Apply latest JS structure from ApplePay Gateway

- Partially addresses the known display bug
- Simplifies maintainance between both gateways
- Reduces component-internal redundancies

stracker-phil committed:
✨ Add new isEligible flag

The flag is set once Google’s PaymentClient responds to the isReadyToPay() request and controls the rendering of the button

stracker-phil committed:
♻️ Move config object to appropriate file

stracker-phil committed:
✨ Sync gateway visibility via custom event

stracker-phil committed:
✨ Introduce new “isVisible” flag

stracker-phil committed:
✨ Introduce a new ConsoleLogger JS class

Extract debug logic to separate component

stracker-phil committed:
✨ New PaymentButton base class for APM buttons

This class is used to render buttons in the front end, and encapsulates logic that is shared between ApplePay and GooglePay buttons

stracker-phil committed:
🚚 Move ConsoleLogger to wc-gateway module

stracker-phil committed:
♻️ Move button event dispatcher to helper file

stracker-phil committed:
♻️ Move most of the display logic to base class

The PaymentButton base class now handles display logic that is shared between different APMs

stracker-phil committed:
🔀 Merge branch 'trunk'

# Conflicts:
#	modules/ppcp-googlepay/resources/js/GooglepayButton.js

stracker-phil committed:
🩹 ConsoleLogger will always output error messages

stracker-phil committed:
🔥 Minor cleanup

stracker-phil committed:
🐛 Fix Google Pay button on non-checkout pages

Reflect changes of PR #2495

stracker-phil committed:
🥅 Improve code stability and documentation

stracker-phil committed:
♻️ Simplify PaymentButton creation code

stracker-phil committed:
🐛 Fix issue with duplicate button instances

On Block Checkout pages, there were multiple PaymentButton instances for the same context.

stracker-phil committed:
🐛 Fix button insertion logic in block checkout

The Google Pay button was inserted into the wrong DOM wrapper

stracker-phil committed:
🐛 Address some errors in preview (Settings page)

stracker-phil committed:
♻️ Unify constructor arguments of PaymentButton

Factory method and constuctor have same arguments in the same order

stracker-phil committed:
💡 Use ConsoleLogger for PreviewButtonManager

Also add some comments and additional log output

stracker-phil committed:
♻️ Extract the GooglePayPreviewButton to own file

Also note the change in line 43 (new file) where we use `new GooglepayButton` instead of `GooglepayButton.createButton`

stracker-phil committed:
🐛 Fix initial visibility of payment button

stracker-phil committed:
⚡️ Only recreate payment button when needed

Check if the button is still in DOM before recreating it

stracker-phil committed:
✨ Provide means to update styling of a button

Before this, we needed to create a new button instance to display changes in button style. This allows updating an existing button instance

stracker-phil committed:
🚧 Allow updating transaction details

stracker-phil committed:
🐛 Fix incorrect preview update on settings page

stracker-phil committed:
🐛 Fix incorrect event listeners

stracker-phil committed:
🐛 Refresh total price when checkout form changes

Applies the selected shipping costs to the Google Payment button

stracker-phil committed:
🐛 Fetch transaction details on button click

Fixes an issue with multiple payment buttons on one page, like Mini-Cart and Single Product

stracker-phil committed:
🐛 Fix remaining issue on classic checkout page

stracker-phil committed:
🔀 Merge branch 'trunk'

# Conflicts:
#	modules/ppcp-googlepay/resources/js/GooglepayButton.js

stracker-phil committed:
🐛 Fix missing button outline in front end

danieldudzic commented:
Great job, really!

One error that I have noticed, is preventing the Google Pay button from loading on the Classic Checkout page:

![google_pay](https://github.com/user-attachments/assets/759f7d95-bd91-4f33-9b1c-673568e0bf61)

danieldudzic commented on modules/ppcp-googlepay/resources/js/GooglepayManager.js:
```diff
@@ -20,7 +20,7 @@ class GooglepayManager {
 			);
 
-			const button = new GooglepayButton(
+			const button = GooglepayButton.createButton(
```
Love this change! 💚

danieldudzic reviewed (approved):
This is golden. Awesome work!

stracker-phil committed:
🔊 Console error when button init fails

stracker-phil committed:
🐛 Fix init errors “No API configuration...”

Caused by calling `reinit()` before the invoking `init()` first. New condition in `reinit()` ensures that the button configuration is present.

stracker-phil committed:
🔥 Remove unused getter

stracker-phil committed:
💡 Improve some code comments

stracker-phil commented:
> ...
> One error that I have noticed, is preventing the Google Pay button from loading on the Classic Checkout page:
> 
> **"No API configuration - missing configure() call?"**

Thanks for reporting

✅ Fixed by this new commit: https://github.com/woocommerce/woocommerce-paypal-payments/pull/2496/commits/966b76102c461e928b8b162ea7f1aaac77ec6582#diff-e0140fc4cf16be408cd39db4c0e6d84cefd19fe55cbe2a2176bd96d9e0fdcc78R289-R291

stracker-phil committed:
💡 Add comments to context options

stracker-phil committed:
🐛 Fix issue with missing button in checkout

When the separate “Google Pay gateway” is disabled, the Google Pay button should be rendered inside the “PayPal gateway” container.

danieldudzic commented:
I reviewed and tested this now again, and everything works as expected 🙌
